clippy_lints/src/boxed_void.rs

```diff
@@ -0,0 +1,50 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::path_def_id;
+use clippy_utils::ty::is_c_void;
+use rustc_hir::{Expr, ExprKind, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::RawPtr;
+use rustc_middle::ty::TypeAndMut;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks if we're passing a `c_void` raw pointer to `Box::from_raw(_)`
+    ///
+    /// ### Why is this bad?
+    /// This is a known pitfall of FFI, particularly with some C or C++ APIs.
```

Comment:
Can you link to some documentation where this pitfall is documented? If you can't find any official documentation, please explain this pitfall in more detail here.

Reply:
The issue's OP made a good description:

> However, it is easy to run into the pitfall of calling `from_raw` with the `c_void` pointer. Note that the definition of, say, `Box::from_raw` is:
>
> ```rust 
> pub unsafe fn from_raw(raw: *mut T) -> Box<T>
> ```
> meaning that if you pass a *mut c_void you will get a `Box<c_void>`. Per the safety requirements in the documentation, for this to be safe, c_void would need to have the same memory layout as the original type, which is often not the case. And even if it was the case, if the original type or one of its fields (or one of its fields, recursively) had a `Drop` implementation, that implementation would not run when this newly recreated box is dropped. (And similarly for `Rc` and `Arc`.)

Reply:
Do I mention the issue link in `Why is this bad?` ?

Reply:
No, please put this explanation into the documentation. This will save Clippy users a click.

Reply:
Is documentation now correct ? Need help

```rust
declare_clippy_lint! {
    /// ### What it does
    /// Checks if we're passing a `c_void` raw pointer to `Box::from_raw(_)`
    ///
    /// ### Why is this bad?
    /// However, it is easy to run into the pitfall of calling from_raw with the c_void pointer.
    /// Note that the definition of, say, Box::from_raw is:
    ///
    /// ```rust
    /// pub unsafe fn from_raw(raw: *mut T) -> Box<T>
    /// ```
    /// meaning that if you pass a *mut c_void you will get a Box<c_void>.
    /// Per the safety requirements in the documentation, for this to be safe,
    /// c_void would need to have the same memory layout as the original type, which is often not the case.
    /// And even if it was the case, if the original type or one of its fields (or one of its fields, recursively) had a Drop implementation, that implementation would not run when this newly recreated box is dropped.
    ///
    /// ### Example
    /// ```rust
    /// #use std::ffi::c_void;
    /// #let ptr = Box::into_raw(Box::new(42usize)) as *mut c_void;
    /// let _ = unsafe { Box::from_raw(ptr) };
    /// ```
    /// Use instead:
    /// ```rust
    /// #use std::ffi::c_void;
    /// #let ptr = Box::into_raw(Box::new(42usize)) as *mut c_void;
    /// let _ = unsafe { Box::from_raw(ptr as *mut usize) };
    /// ```
    ///
    #[clippy::version = "1.66.0"]
    pub RAW_BOX_FROM_VOID_PTR,
    suspicious,
    "creating a `Box` from a raw void pointer"
}
```

Reply:
I would leave out the `And even if it was the case, ...` line. Otherwise LGTM